lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -396,21 +379,326 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+    // accessing the tiles in non-row-major order.
+    //
+    // iterators are just ints that correspond to absolute offsets in the CB.
+    // They walk the order defined by the affine map associated with the memref.
+    LoopNest loopNest;
+    loopNest.blockArgIteratorMapping = blockArgIteratorMapping;
+    SmallVector<scf::ForOp> loops;
+    SmallVector<Region *> loopRegions;
+    SmallVector<SmallVector<Value>> iteratorsNest = {uniqueIterators};
+    for (unsigned dim = 0; dim < outputAffineMap.getNumDims(); ++dim) {
+      OpBuilder regionBuilder(builder);
+      if (!loopNest.loopRegions.empty()) {
+        regionBuilder = OpBuilder(loopNest.loopRegions.back());
+      }
+      // Loop variables, these are decoupled from the iterators
+      Value lowerBound = i32(0, regionBuilder);
+      Value upperBound = i32(outputShape[dim], regionBuilder);
+      Value loopStep = i32(1, regionBuilder);
+      scf::ForOp forOp = regionBuilder.create<scf::ForOp>(
+          output.getLoc(), lowerBound, upperBound, loopStep,
+          iteratorsNest.back());
+      loopNest.loops.push_back(forOp);
+
+      SmallVector<std::int64_t> innerIndexStep(outputAffineMap.getNumDims(), 0);
+      innerIndexStep[dim] = 1;
+
+      bool innerLoop = dim == (outputAffineMap.getNumDims() - 1);
+      if (innerLoop) {
+        OpBuilder innerLoopRegion(loopNest.loops.back().getRegion());
+        SmallVector<Value> innerIndices;
+        int i = 0;
+        for (auto [affineMap, iterator] : iteratorMaps) {
+          // Calculate how far a single step in the inner dim is.
+          SmallVector<std::int64_t> innerOffset =
+              affineMap.compose(innerIndexStep);
+          assert(innerOffset.size() == 1);
+          innerIndices.push_back(innerLoopRegion.create<arith::AddIOp>(
+              output.getLoc(), forOp.getRegionIterArg(i),
+              i32(innerOffset[0], innerLoopRegion)));
+          ++i;
+        }
+        innerLoopRegion.create<scf::YieldOp>(output.getLoc(), innerIndices);
+      }
+
+      // Backpedal and adjust the iterator offset for the current dimension.
+      if (dim > 0) {
+        SmallVector<Value> outerIndices;
+        SmallVector<std::int64_t> outerIndexStep(outputAffineMap.getNumDims(),
+                                                 0);
+        outerIndexStep[dim - 1] = 1;
+        int i = 0;
+        for (auto [affineMap, iterator] : iteratorMaps) {
+          // Calculate how far a single step in the inner dim is.
+          SmallVector<std::int64_t> innerOffset =
+              affineMap.compose(innerIndexStep);
+          assert(innerOffset.size() == 1);
+          // Calculate how far a single step in the outer dim is.
+          SmallVector<std::int64_t> outerOffset =
+              affineMap.compose(outerIndexStep);
+          assert(outerOffset.size() == 1);
+          // Multiply by the number of steps that the inner loop took.
+          // FIXME: test this for higher dims
+          std::int64_t offset =
```

Comment:
Do you subtract `innerOffset` because scf.for updates all iterators regardless of nesting level, and therefore `outerOffset` is incremented unnecessary?

Reply:
This, theoretically, should handle cases with arbitrary walk order.  For example, the memref's affine map might walk the input buffer in transposed order, so let's say we have shard in tiles `3x4`.  Since the shard is stored row major, the inner loops that index this transposed would look like:
```c++
int iter = 0;
for (int x = 0; x < 4; ++y) {
    for (int y = 0; y < 3; ++y) {
      unpack_tile(iter)
      iter += 4;
    }
    // outerOffset = 1
    // innerOffset = 4
    // outputShape[dim] = 3
    iter = iter + (1 - 4 * 3);
}
```

The goal is to let the kernel implement arbitrary walk orders so that we can do less data movement using noc.